src/platform/Linux/MdnsImpl.h

```diff
@@ -0,0 +1,167 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <sys/select.h>
+#include <unistd.h>
+
+#include <chrono>
+#include <map>
+#include <memory>
+#include <set>
+#include <vector>
+
+#include <avahi-client/client.h>
+#include <avahi-client/lookup.h>
+#include <avahi-client/publish.h>
+#include <avahi-common/domain.h>
+#include <avahi-common/error.h>
+#include <avahi-common/watch.h>
+
+#include "lib/protocols/mdns/Mdns.h"
+
+struct AvahiWatch
+{
+    int mFd;                      ///< The file descriptor to watch.
+    AvahiWatchEvent mWatchEvents; ///< The interested events.
+    int mHappenedEvents;          ///< The events happened.
+    AvahiWatchCallback mCallback; ///< The function to be called when interested events happened on mFd.
+    void * mContext;              ///< A pointer to application-specific context.
+    void * mPoller;               ///< The poller created this watch.
+};
+
+struct AvahiTimeout
+{
+    std::chrono::steady_clock::time_point mAbsTimeout; ///< Absolute time when this timer timeout.
+    AvahiTimeoutCallback mCallback;                    ///< The function to be called when timeout.
+    bool mEnabled;                                     ///< Whether the timeout is enabled.
+    void * mContext;                                   ///< The pointer to application-specific context.
+    void * mPoller;                                    ///< The poller created this timer.
+};
+
+namespace chip {
+namespace Protocols {
+namespace Mdns {
+
+class Poller
+{
+public:
+    Poller(void);
+
+    void UpdateFdSet(fd_set & readFdSet, fd_set & writeFdSet, fd_set & errorFdSet, int & maxFd, timeval & timeout);
+
+    void Process(const fd_set & readFdSet, const fd_set & writeFdSet, const fd_set & errorFdSet);
+
+    const AvahiPoll * GetAvahiPoll(void) const { return &mAvahiPoller; }
+
+private:
+    static AvahiWatch * WatchNew(const struct AvahiPoll * poller, int fd, AvahiWatchEvent event, AvahiWatchCallback callback,
```

Comment:
Please move this static function to the .cpp file in an unnamed namespace.

Reply:
These functions need to call private member functions (`WatchNew` without the context ptr) so I declare it as a class member. Declaring as friend doesn't show much advantage.

Reply:
I don't agree with a style where callbacks or virtual overrides have private visibility when they can be called without restriction by another module (in this case avahi). They should be public.